File: src/com/suushiemaniac/cubing/bld/model/enumeration/puzzle/CubicPuzzle.java
package com.suushiemaniac.cubing.bld.model.enumeration.puzzle;

import com.suushiemaniac.cubing.alglib.lang.CubicAlgorithmReader;
import com.suushiemaniac.cubing.alglib.lang.NotationReader;
import com.suushiemaniac.cubing.bld.analyze.*;
import com.suushiemaniac.cubing.bld.model.enumeration.piece.PieceType;
import com.suushiemaniac.cubing.bld.util.ClosureUtil;
import net.gnehzr.tnoodle.scrambles.Puzzle;
import puzzle.*;

import java.util.Collections;
import java.util.function.Function;
import java.util.function.Supplier;

import static com.suushiemaniac.cubing.bld.model.enumeration.piece.CubicPieceType.*;

public enum CubicPuzzle implements TwistyPuzzle {
	TWO(2, TwoByTwoCubePuzzle::new, CORNER),
	THREE(3, ThreeByThreeCubePuzzle::new, CORNER, CENTER, EDGE),
	THREE_BLD(NoInspectionThreeByThreeCubePuzzle::new, THREE),
	THREE_FMC(ThreeByThreeCubeFewestMovesPuzzle::new, THREE),
	FOUR(4, FourByFourCubePuzzle::new, CORNER, XCENTER, WING),
	FOUR_RAND(FourByFourRandomTurnsCubePuzzle::new, FOUR),
	FOUR_BLD(NoInspectionFourByFourCubePuzzle::new, FOUR),
	FIVE(5, CubePuzzle::new, CORNER, CENTER, EDGE, XCENTER, TCENTER, WING),
	FIVE_BLD(NoInspectionFiveByFiveCubePuzzle::new, FIVE),
	SIX(6, CubePuzzle::new, CORNER, XCENTER, WING, INNERXCENTER, INNERWING, LEFTOBLIQUE, RIGHTOBLIQUE),
	SEVEN(7, CubePuzzle::new, CORNER, CENTER, EDGE, XCENTER, TCENTER, WING, INNERXCENTER, INNERWING, LEFTOBLIQUE, RIGHTOBLIQUE, INNERTCENTER);

	public static CubicPuzzle fromSize(int size, String flag) {
		for (CubicPuzzle puzzle : values()) {
			if (puzzle.getSize() == size) {
				if (flag.length() <= 0) {
					if (!puzzle.name().contains("_")) {
						return puzzle;
					}
				} else if (puzzle.name().toUpperCase().endsWith(flag.toUpperCase())) {
					return puzzle;
				}
			}
		}

		if (size >= 2 && size <= 7) {
			return fromSize(size);
		}

		return THREE;
	}

	public static CubicPuzzle fromSize(int size) {
		return fromSize(size, "");
	}

	private static NotationReader READER_INST = new CubicAlgorithmReader();

	private int size;
	private Supplier<Puzzle> scramblingPuzzleGen;
	private PieceType[] types;

	private BldPuzzle analyzingPuzzle;

	CubicPuzzle(int size, Supplier<Puzzle> scramblingPuzzleGen, PieceType... types) {
		this.size = size;
		this.scramblingPuzzleGen = scramblingPuzzleGen;
		this.types = types;

		this.analyzingPuzzle = new BldCube(this);
	}

	CubicPuzzle(int size, Function<Integer, Puzzle> scramblingPuzzleGen, PieceType... types) {
		this(size, () -> scramblingPuzzleGen.apply(size), types);
	}

	CubicPuzzle(Supplier<Puzzle> scramblingPuzzleGen, CubicPuzzle parent) {
		this(parent.getSize(), scramblingPuzzleGen, parent.getPieceTypes());
	}

	@Override
	public int getSize() {
		return this.size;
	}

	@Override
	public Supplier<Puzzle> supplyScramblingPuzzle() {
		return this.scramblingPuzzleGen;
	}

	@Override
	public BldPuzzle getAnalyzingPuzzle() {
		return this.analyzingPuzzle.clone();
	}

	@Override
	public PieceType[] getPieceTypes() {
		return this.types;
	}

	@Override
	public NotationReader getReader() {
		return READER_INST;
	}

	@Override
	public String toString() {
		return String.join("x", Collections.nCopies(3, String.valueOf(this.size)));
	}
}

File: src/main/kotlin/com/suushiemaniac/cubing/bld/model/enumeration/puzzle/CubicPuzzle.kt
package com.suushiemaniac.cubing.bld.model.enumeration.puzzle

import com.suushiemaniac.cubing.alglib.alg.Algorithm
import com.suushiemaniac.cubing.alglib.lang.CubicAlgorithmReader
import com.suushiemaniac.cubing.alglib.lang.NotationReader
import com.suushiemaniac.cubing.bld.gsolve.KPuzzle
import com.suushiemaniac.cubing.bld.model.enumeration.piece.PieceType
import com.suushiemaniac.cubing.bld.util.StringUtil.repeatWithGap
import net.gnehzr.tnoodle.scrambles.Puzzle
import puzzle.*

import com.suushiemaniac.cubing.bld.model.enumeration.piece.CubicPieceType.*
import java.io.File

enum class CubicPuzzle(override val size: Int, val scramblingPuzzleGen: () -> Puzzle, vararg types: PieceType) : TwistyPuzzle {
    TWO(2, ::TwoByTwoCubePuzzle, CORNER),
    THREE(3, ::ThreeByThreeCubePuzzle, CORNER, CENTER, EDGE),
    THREE_BLD(::NoInspectionThreeByThreeCubePuzzle, THREE),
    THREE_FMC(::ThreeByThreeCubeFewestMovesPuzzle, THREE),
    FOUR(4, ::FourByFourCubePuzzle, CORNER, XCENTER, WING),
    FOUR_RAND(::FourByFourRandomTurnsCubePuzzle, FOUR),
    FOUR_BLD(::NoInspectionFourByFourCubePuzzle, FOUR),
    FIVE(5, ::CubePuzzle, CORNER, CENTER, EDGE, XCENTER, TCENTER, WING),
    FIVE_BLD(::NoInspectionFiveByFiveCubePuzzle, FIVE),
    SIX(6, ::CubePuzzle, CORNER, XCENTER, WING, INNERXCENTER, INNERWING, LEFTOBLIQUE, RIGHTOBLIQUE),
    SEVEN(7, ::CubePuzzle, CORNER, CENTER, EDGE, XCENTER, TCENTER, WING, INNERXCENTER, INNERWING, LEFTOBLIQUE, RIGHTOBLIQUE, INNERTCENTER);

    constructor(size: Int, scramblingPuzzleGen: (Int) -> Puzzle, vararg types: PieceType) : this(size, { -> scramblingPuzzleGen(size) }, *types)
    constructor(scramblingPuzzleGen: () -> Puzzle, parent: CubicPuzzle) : this(parent.size, scramblingPuzzleGen, *parent.pieceTypes)

    override val pieceTypes: Array<out PieceType> = types

    override val reader: NotationReader
        get() = READER_INST

    override val kPuzzle: KPuzzle // FIXME not load a new kPuzzle every time at get
        get() = KPuzzle(File(this.javaClass.getResource("permutations/${this.size.toString().repeat(3)}.def").toURI()))

    override val scramblingPuzzle: Puzzle = this.scramblingPuzzleGen()

    override fun independentScrambleSupplier(): Sequence<Algorithm> {
        val independentTNoodle = this.scramblingPuzzleGen()
        val independentReader = getReader()

        return generateSequence {
            independentReader.parse(independentTNoodle.generateScramble())
        }
    }

    override fun toString(): String {
        return this.size.toString().repeatWithGap(3, "x")
    }

    companion object {
        fun fromSize(size: Int, flag: String = ""): CubicPuzzle {
            for (puzzle in values()) {
                if (puzzle.size == size) {
                    if (flag.isEmpty()) {
                        if (!puzzle.name.contains("_")) {
                            return puzzle
                        }
                    } else if (puzzle.name.toUpperCase().endsWith(flag.toUpperCase())) {
                        return puzzle
                    }
                }
            }

            return if (size in 2..7) {
                fromSize(size)
            } else THREE

        }

        fun getReader(): NotationReader {
            return CubicAlgorithmReader()
        }

        private val READER_INST = getReader()
    }
}
